File: 3_find_disappeared_numbers/3.py
from typing import List

# https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/

class Solution:
    # O(n) time complexity, O(n) space complexity
    def findDisappearedNumbers_1(self, nums: List[int]) -> List[int]:
        unique_numbers = set(nums)
        missing_numbers = []

        for num in range(1, len(nums) + 1):
            if num not in unique_numbers:
                missing_numbers.append(num)

        return missing_numbers

    # O(n) time complexity, O(1) space complexity
    # подсмотрено на ytube, разобрано
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        # По условию числа от 1 до n, а индексы от 0 до n-1.
        # Для каждого числа num пометим индекс num-1 с помощью отрицания числа
        for num in nums:
            i = abs(num) - 1
            if nums[i] > 0:
                nums[i] *= -1

        # Обратное преобразование.
        # Положительный num означает, что элемент равный i+1 не был в исходном массиве
        missing = []
        for i, num in enumerate(nums):
            if num > 0:
                missing.append(i + 1)

        return missing

File: 3_find_disappeared_numbers/3.go
package find_disappeared_numbers

// https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/

// O(n) time complexity, O(n) space complexity
func findDisappearedNumbers2(nums []int) []int {
	unique := make(map[int]bool)
	for i := 1; i <= len(nums); i++ {
		unique[i] = true
	}

	for _, num := range nums {
		if unique[num] {
			delete(unique, num)
		}
	}

	disappeared := make([]int, len(unique))
	i := 0
	for num := range unique {
		disappeared[i] = num
		i++
	}

	return disappeared
}

// O(n) time complexity, O(1) space complexity
func findDisappearedNumbers(nums []int) []int {
	// По условию числа от 1 до n, а индексы от 0 до n-1.
	// Для каждого числа num пометим индекс num-1 с помощью отрицания числа
	for i := 0; i < len(nums); i++ {
		num := nums[i]
		index := Abs(num) - 1

		if nums[index] > 0 {
			nums[index] *= -1
		}
	}

	// Обратное преобразование.
	// Положительный num означает, что элемент равный i+1 не был в исходном массиве
	var disappeared []int
	for i, num := range nums {
		if num > 0 {
			disappeared = append(disappeared, i+1)
		}
	}

	return disappeared
}

func Abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}
